### ISSUE TITLE
Specify environment variables when declaring a rule

### ISSUE THREAD

**Issue by wardellbagby** [bazelbuild/rules_nodejs#2261]
# 🚀 feature request

### Relevant Rules

- npm_package_bin
- nodejs_binary
- [Generated macros for npm packages with `bin` entries](https://bazelbuild.github.io/rules_nodejs/repositories.html#generated-macros-for-npm-packages-with-bin-entries)

### Description

The ability to set environment variables on rules that will be set in the shell when the rule executes.

### Describe the solution you'd like

My Bazel lingo isn't great, so my description might not be using the right terminology, but my preferred solution would look like this:

```
npm_package_bin(
    name = "bundle",
    args = [
        "--config", "my/config/file.js"
        ],
    data = [
        "@npm//webpack-cli",
    ],
    env = {
        MY_ENV_VARIABLE = "some_value"
    },
    tool = "@npm//webpack/bin:webpack",
    output_dir = True,
    visibility = ["//visibility:public"]
)
```

This is great for tools that require certain information to be set as an environment variable (in my case, the `TS_NODE_PROJECT` that Webpack wants to read) without potentially polluting any other rules that either don't care about that environment variable or need to read a different value for it.

### Describe alternatives you've considered

- `configuration_env_vars` / `default_env_vars`
   - Requires setting the env variable either when running the command (e.g.,`bazel run //path:my_rule --action_env=MY_ENV_VARIABLE=some_value)` or in the `.bazelrc`. This would be fine but I only need the environment variable to be set when the rule is executed and I'll have multiple rules that'll need to read the same environment variable that would need to read different values for it. 
 - [Dropbox's Node Bazel rules](https://github.com/dropbox/rules_node#webpack_build)
    - This does exactly what I want, but is currently unsupported and can't be used with the generated npm repository from `rules_nodejs` since it doesn't do the NPM dependency linking.
 - `genrule`
    - Would let me just do `export MY_ENV_VARIABLE=some_value` before calling whatever script I wanted, but also doesn't do the NPM dependency linking.

**Comment by wardellbagby** [comment]
Not stale!

**Comment by KatSick** [comment]
If there any updates? I see in bazel 3.3.0 we now can specify [env](https://bazelbuild.github.io/rules_nodejs/Built-ins.html#nodejs_binary-env) variable, however, it still does not work with webpack + TS_NODE_PROJECT

**Comment by mattem** [comment]
@KatSick are you able to provide a repro demonstrating the failure case?

**Comment by KatSick** [comment]
Hi @mattem !

During creating the minimal repo, I found how to enable TS_NODE_PROJECT to work.
I missing `execpath` in env variable.

So, here is my BUILD.bazel config that works with **webpack@5**.
For webpack@4, you need to add also `-r ts-node/register` in the args (to excplicitly enable ts-node)

Also, for whom it may concern, here is fully working repo: https://github.com/KatSick/bazel-webpack-ts-example

```load("@npm//webpack-cli:index.bzl", webpack = "webpack_cli")

webpack(
    name = "bazel-ts_node_project-repro",
    outs = ["dist"],
    args = [
        "--config",
        "$(execpath webpack.config.ts)",
        "--output-path",
        "$@",
    ],
    data = glob([
        "src/**/*",
        "webpack.config*",
        "tsconfig*",
    ]) + [
        "@npm//webpack",
        "@npm//ts-loader",
        "@npm//typescript",
        "@npm//tslib",
        "@npm//clean-webpack-plugin",
        "@npm//fork-ts-checker-webpack-plugin",
        "@npm//html-webpack-plugin",
        "@npm//mini-css-extract-plugin",
        "@npm//@types/css-minimizer-webpack-plugin",
        "@npm//@types/node",
        "@npm//@types/webpack",
        "@npm//@types/mini-css-extract-plugin",
        "@npm//css-minimizer-webpack-plugin",
        "@npm//terser-webpack-plugin",
        "@npm//ts-node",
        "@npm//tsconfig-paths",
        "@npm//webpack-cli",
        "@npm//webpack-dev-server",
        "@npm//webpack-merge",
    ],
    env = {"TS_NODE_PROJECT": "$(execpath tsconfig.webpack.json)"},
)
```

**Comment by mattem** [comment]
Thanks for digging into that!

I'm going to close this as the `env` attribute is available as of 3.3.0
